src/ralph/schemas/edx/base.py

```diff
@@ -0,0 +1,144 @@
+            raise ValidationError("org_id should be empty if course_id is empty")
+        if data["org_id"] == "":
+            return
+        if not data["course_id"].startswith("course-v1:"):
+            raise ValidationError("course_id should starts with 'course-v1'")
+        try:
+            organization, course, session = (
+                data["course_id"].replace("course-v1:", "").split("+")
+            )
+        except ValueError:
+            raise ValidationError(
+                "course_id should contain an organization ID, "
+                "a course name and session separated by a +"
+            )
+        if organization != data["org_id"]:
+            raise ValidationError(
+                "organization ID in the course ID does not match the event organization ID"
+            )
+        if len(course) == 0 or len(session) == 0:
+            raise ValidationError("course and session should not be empty")
+
+
+class ContextModuleSchema(Schema):
+    """Represents the context module field"""
+
+    usage_key = fields.Str(required=True)
+    display_name = fields.Str(required=True)
+
+
+class ContextSchema(BaseContextSchema):
+    """Context with module field. Present in Capa problems related
+    events.
+    """
+
+    module = fields.Nested(ContextModuleSchema(), required=True)
+
+    # pylint: disable=no-self-use
+    @validates_schema
+    def validate_path(self, data, **kwargs):
+        """path should be equal to
+        "/courses/{course_id}/xblock/block-v1:{course_id[10:]}"
+        "+type@problem+block@{usage_key[-32:]}"
+        "/handler/xmodule_handler/problem_check"
+        """
+        if data["path"] != (
+            f"/courses/"
+            f"{data['course_id']}"
+            f"/xblock/"
+            f"{data['module']['usage_key']}/"
+            f"handler/xmodule_handler/problem_check"
+        ):
+            raise ValidationError(
+                "path should be equal to"
+                "/courses/{course_id}/xblock/block-v1:{course_id[10:]}"
+                "+type@problem+block@{usage_key[-32:]}"
+                "/handler/xmodule_handler/problem_check"
+            )
+
+
+class IPAddressField(fields.Field):
+    """IP Address that serializes to a string of numbers and deserializes
+    to a IPv4Address object.
+    """
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        if value == "":
+            return ""
+        return value.exploded
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        if value == "":
+            return ""
+        try:
+            return IPv4Address(value)
+        except ValueError as error:
+            raise ValidationError("Invalid IPv4 Address") from error
+
+
+def validate_username(value):
+    """"check username field empty or 2-30 chars long"""
+    if len(value) == 1 or len(value) > 30:
+        raise ValidationError("username should be empty or between 2 and 30 chars long")
```

Comment:
Are you sure about this? Is this configurable in openedx?

Reply:
I prefer the following (extract from `FeedbackDisplayedSchema`):

```python
username = fields.Str(required=True, validate=Length(min=2, max=30))
```

Reply:
The username can be empty if the user is not connected (for server events)

Reply:
feedback_displayed events are triggered from pages containing capa problems - for course pages the user have to be connected - so the username can't be empty - if I understood is right

Reply:
So what about the following? Will this field be validated only if provided?

```python
username = fields.Str(required=False, validate=Length(min=2, max=30))
```

Reply:
I think yes, but in our case the field is always provided.
I will give it a check, thanks

Reply:
Unfortunately this seems to not be working when:
"username": ""